ext/standard/http.c

```diff
@@ -18,19 +18,215 @@
+	const char *num_prefix, size_t num_prefix_len,
+	const zend_string *key_prefix
+) {
+	zend_string *new_prefix;
+
+	if (index_string) {
+		zend_string *encoded_key;
+		if (encoding_type == PHP_QUERY_RFC3986) {
+			encoded_key = php_raw_url_encode(index_string, index_string_len);
+		} else {
+			encoded_key = php_url_encode(index_string, index_string_len);
+		}
+
+		if (key_prefix) {
+			new_prefix = zend_string_concat3(ZSTR_VAL(key_prefix), ZSTR_LEN(key_prefix), ZSTR_VAL(encoded_key), ZSTR_LEN(encoded_key), "%5D%5B", strlen("%5D%5B"));
+		} else {
+			new_prefix = zend_string_concat2(ZSTR_VAL(encoded_key), ZSTR_LEN(encoded_key), "%5B", strlen("%5B"));
+		}
+		zend_string_release_ex(encoded_key, false);
+	} else { /* is integer index */
+		char *index_int_as_str;
+		size_t index_int_as_str_len;
+
+		index_int_as_str_len = spprintf(&index_int_as_str, 0, ZEND_LONG_FMT, index_int);
+
+		if (key_prefix && num_prefix) {
+			/* zend_string_concat4() */
+			size_t len = ZSTR_LEN(key_prefix) + num_prefix_len + index_int_as_str_len + strlen("%5D%5B");
+			new_prefix = zend_string_alloc(len, 0);
+
+			memcpy(ZSTR_VAL(new_prefix), ZSTR_VAL(key_prefix), ZSTR_LEN(key_prefix));
+			memcpy(ZSTR_VAL(new_prefix) + ZSTR_LEN(key_prefix), num_prefix, num_prefix_len);
+			memcpy(ZSTR_VAL(new_prefix) + ZSTR_LEN(key_prefix) + num_prefix_len, index_int_as_str, index_int_as_str_len);
+			memcpy(ZSTR_VAL(new_prefix) + ZSTR_LEN(key_prefix) + num_prefix_len +index_int_as_str_len, "%5D%5B", strlen("%5D%5B"));
+			ZSTR_VAL(new_prefix)[len] = '\0';
+		} else if (key_prefix) {
+			new_prefix = zend_string_concat3(ZSTR_VAL(key_prefix), ZSTR_LEN(key_prefix), index_int_as_str, index_int_as_str_len, "%5D%5B", strlen("%5D%5B"));
+		} else if (num_prefix) {
+			new_prefix = zend_string_concat3(num_prefix, num_prefix_len, index_int_as_str, index_int_as_str_len, "%5B", strlen("%5B"));
+		} else {
+			new_prefix = zend_string_concat2(index_int_as_str, index_int_as_str_len, "%5B", strlen("%5B"));
+		}
+		efree(index_int_as_str);
+	}
+
+	return new_prefix;
+}
+
+static void php_url_encode_object(zend_object *object, smart_str *form_str,
+	int encoding_type, zend_ulong index_int,
+	const char *index_string, size_t index_string_len,
+	const char *num_prefix, size_t num_prefix_len,
+	const zend_string *key_prefix,
+	const zend_string *arg_sep)
+{
+	if (GC_IS_RECURSIVE(object)) {
+		/* Prevent recursion */
+		return;
+	}
+
+	HashTable *properties = php_object_get_visible_properties(object);
+	if (zend_hash_num_elements(properties) == 0) {
+		zend_array_destroy(properties);
+
+		zval tmp;
+		/* If the data object is stringable without visible properties handle it like a string instead of empty array */
+		if (object->handlers->cast_object(object, &tmp, IS_STRING) == SUCCESS) {
+			php_url_encode_scalar(&tmp, form_str,
+				encoding_type, index_int,
+				index_string, index_string_len,
+				num_prefix, num_prefix_len,
+				NULL,
+				arg_sep);
```

Comment:
Don't forget to destroy tmp after the php_url_encode_scalar call: if __toString returns a non-interned string then tmp could leak.